fem/libceed/ceedsolvers-interpolation.cpp

```diff
@@ -0,0 +1,163 @@
+                             &fine_r_data); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(c_fine_multiplicity, CEED_MEM_HOST,
+                                 &fine_data); CeedChk(ierr);
+   for (int i = 0; i < height; ++i)
+   {
+      fine_r_data[i] = 1.0 / fine_data[i];
+   }
+
+   ierr = CeedVectorRestoreArray(fine_multiplicity_r, &fine_r_data); CeedChk(ierr);
+   ierr = CeedVectorRestoreArrayRead(c_fine_multiplicity, &fine_data);
+   CeedChk(ierr);
+   ierr = CeedVectorDestroy(&c_fine_multiplicity); CeedChk(ierr);
+
+   CeedVector fine_work;
+   ierr = CeedVectorCreate(ceed, height, &fine_work); CeedChk(ierr);
+
+   // ierr = CeedCalloc(1, interp);
+   *interp = (CeedInterpolation) calloc(1,
+                                        sizeof(struct CeedInterpolation_private));
+   (*interp)->ceed = ceed;
+   (*interp)->qf_restrict = qf_restrict;
+   (*interp)->qf_prolong = qf_prolong;
+   (*interp)->op_interp = op_interp;
+   (*interp)->op_restrict = op_restrict;
+   (*interp)->fine_multiplicity_r = fine_multiplicity_r;
+   (*interp)->fine_work = fine_work;
+
+   return 0;
+}
+
+/// this is not implemented with reference counting etc.
+int CeedInterpolationDestroy(CeedInterpolation *interp)
+{
+   int ierr;
+
+   ierr = CeedQFunctionDestroy(&(*interp)->qf_restrict); CeedChk(ierr);
+   ierr = CeedQFunctionDestroy(&(*interp)->qf_prolong); CeedChk(ierr);
+   ierr = CeedOperatorDestroy(&(*interp)->op_interp); CeedChk(ierr);
+   ierr = CeedOperatorDestroy(&(*interp)->op_restrict); CeedChk(ierr);
+   ierr = CeedVectorDestroy(&(*interp)->fine_multiplicity_r); CeedChk(ierr);
+   ierr = CeedVectorDestroy(&(*interp)->fine_work); CeedChk(ierr);
+
+   free(*interp);
+   return 0;
+}
+
+/// @todo could use a CEED_REQUEST here
+int CeedInterpolationInterpolate(CeedInterpolation interp,
+                                 CeedVector in, CeedVector out)
+{
+   int ierr;
+
+   ierr = CeedOperatorApply(interp->op_interp, in, out,
+                            CEED_REQUEST_IMMEDIATE); CeedChk(ierr);
+   ierr = CeedVectorPointwiseMult(out, interp->fine_multiplicity_r); CeedChk(ierr);
+
+   return 0;
+}
+
+/// @todo could use a CEED_REQUEST here
+/// @todo using MFEM_FORALL in this Ceed-like function is ugly
+int CeedInterpolationRestrict(CeedInterpolation interp,
+                              CeedVector in, CeedVector out)
+{
+   int ierr;
+
+   int length;
+   ierr = CeedVectorGetLength(in, &length); CeedChk(ierr);
+
+   const CeedScalar *multiplicitydata, *indata;
+   CeedScalar *workdata;
+   CeedMemType mem;
+   CeedGetPreferredMemType(interp->ceed, &mem);
+   ierr = CeedVectorGetArrayRead(in, mem, &indata); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(interp->fine_multiplicity_r, mem,
+                                 &multiplicitydata); CeedChk(ierr);
+   ierr = CeedVectorGetArray(interp->fine_work, mem, &workdata); CeedChk(ierr);
+   MFEM_FORALL(i, length,
+   {workdata[i] = indata[i] * multiplicitydata[i];});
```

Comment:
`mem` does not have the right value, you need to set `mem` according the mode of `MFEM_FORALL`.

Reply:
Done.